src/sync/spsc_bufferqueue.rs

```diff
@@ -0,0 +1,400 @@
+
+            _dummy_2: CachePadded::zeroed(),
+            tail: AtomicUsize::new(0),
+            head_cache: AtomicUsize::new(0),
+            cons_alive: AtomicBool::new(true),
+        };
+        let qarc = Arc::new(q);
+        let rtuple = (BufferProducer::new(qarc.clone()),
+                      BufferConsumer::new(qarc));
+        fence(SeqCst);
+        rtuple
+    }
+
+    /// Tries constructing the element and inserts into the queue
+    ///
+    /// Returns the closure if there isn't space
+    #[inline(always)]
+    fn try_construct<F>(&self, ctor: F) -> Result<(), F> where F: FnOnce() -> T {
+        let ctail = self.tail.load(Relaxed);
+        let mut next_tail = ctail + 1;
+        next_tail = if next_tail == self.size  { 0 } else { next_tail };
+        if next_tail == self.head_cache.load(Relaxed) {
+            let cur_head = self.head.load(Acquire);
+            self.head_cache.store(cur_head, Relaxed);
+            if next_tail == cur_head {
+                return Err(ctor);
+            }
+        }
+        unsafe {
+            let data_pos = self.data_block.offset(ctail as isize);
+            ptr::write(data_pos, ctor());
+        }
+        self.tail.store(next_tail, Release);
+        Ok(())
+    }
+
+    pub fn try_pop(&self) -> Option<T> {
+        let chead = self.head.load(Relaxed);
+        if chead == self.tail_cache.load(Relaxed) {
+            let cur_tail = self.tail.load(Acquire);
+            self.tail_cache.store(cur_tail, Relaxed);
+            if chead == cur_tail {
+                return None;
+            }
+        }
+
+        let mut next_head = chead + 1;
+        next_head = if next_head == self.size  { 0 } else { next_head };
+        unsafe {
+            let data_pos = self.data_block.offset(chead as isize);
+            let rval = Some(ptr::read(data_pos));
+            self.head.store(next_head, Release);
+            rval
+        }
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.size - 1 //extra space added in ctor as buffer for head/tail
+    }
+}
+
+
+impl<T: Send> Drop for SpscBufferQueue<T> {
+    fn drop(&mut self) {
+        fence(SeqCst);
+        loop {
+            if let None = self.try_pop() {
+                break;
+            }
+        }
+        unsafe { deallocate(self.data_block, self.capacity()); }
+    }
+}
+
+/// The consumer proxy for the SpscBufferQueue
+pub struct BufferConsumer<T: Send> {
+    spsc: Arc<SpscBufferQueue<T>>,
+}
+
+unsafe impl<T: Send> Send for BufferConsumer<T> {}
+
+impl<T: Send> Drop for BufferConsumer<T> {
+    fn drop(&mut self) {
+        self.spsc.cons_alive.store(false, Release);
+    }
+}
+
+impl<T: Send> BufferConsumer<T> {
+    pub fn new(queue: Arc<SpscBufferQueue<T>>) -> BufferConsumer<T> {
+        BufferConsumer {
+            spsc: queue,
+        }
+    }
+
+    /// Creates a new producer if the current one is dead
+    pub fn create_producer(&self) -> Option<BufferProducer<T>> {
```

Comment:
Hmm -- I don't quite understand the motivation for this API design. Why not follow the design of Rust's standard channels, and simply provide a free function in the module that generates a producer/consumer pair?